design/reboot-interface.md

```diff
@@ -0,0 +1,190 @@
+controller that they are ready to continue by removing the annotation with their 
+unique ``{key}`` suffix.
+
+In the case of multiple clients, the controller will wait for all annotations 
+of the form ``reboot.metal3.io/{key}`` to be removed before powering on the Host.
+
+If both ``reboot.metal3.io`` and ``reboot.metal3.io/{key}`` forms are in use, 
+the  ``reboot.metal3.io/{key}`` form will take priority.
+
+In all cases, the content of the annotation is ignored but preserved. This
+ensures that whatever placed the annotation can have a way of tracing it back
+to its source. In the case of remediation, the content will be the UID of the
+Machine resource being remediated.
+
+The actual power management will be performed by the Host controller. This is
+necessary to avoid race conditions by ensuring that the ``Online`` flag and any
+reboot requests are managed in the same place. 
+
+If:
+
+* there is one or more annotations with the ``reboot.metal3.io`` prefix
+  present, and
+* the ``Status.PoweredOn`` field is true, and
+* the value of ``pendingRebootSince`` is empty or earlier than the
+  ``lastPoweredOn`` time
+
+then the Host controller will update ``pendingRebootSince`` to the current
+time.
+
+Whenever ``pendingRebootSince`` is later than the ``lastPoweredOn`` time, the
+Host controller will attempt to power the host off regardless of the
+``Spec.Online`` setting.
+
+Once the Host is powered off ( ``Status.PoweredOn`` is false ), if/when
+
+* the ``Spec.Online`` field is true, and
+* the ``lastPoweredOn`` time is before the ``pendingRebootSince`` time
+
+then the controller should remove the suffixless ``reboot.metal3.io``
+annotation (if present). Once no reboot annotations  are present (i.e. those of
+the form ``reboot.metal3.io/{key}`` have been removed by their originators),
+the existing logic for powering on the Host should execute and update the
+``lastPoweredOn`` timestamp accordingly.
+
+The controller automatically removes all annotations with the 
+``reboot.metal3.io`` prefix if
+* the Host is deprovisioned
+
+## Drawbacks
+
+## Future Enhancements
+
+### Defining a Formal User Interface
+
+
+### Soft Reboots
+
+The resource could accept a "soft" reboot option, in which we would attempt to
+reboot the Host without switching the power off. Since these requests may fail,
+more error reporting would likely be required. In the event that a mix of hard
+and soft reboot requests were pending, the host would be hard-rebooted and all
+of the pending requests marked complete.
+
+## Alternatives
+
+The Machine (Reboot) Actuator could perform the reboot in an imperative, rather
+than declarative, manner by manipulating the ``Online`` spec of the Host in
+sequence to ``false`` and then back to ``true`` again once the ``poweredOn``
+status was seen to be ``false``. However, in the presence of multiple actors
+this approach is prone to race conditions or controller fights.
+
```

Comment:
```suggestion

We could implement a ``HostOn`` and/or ``HostOff`` CRD with the same name as
the target Machine.  This would:
* ensure uniqueness
* move all power management into a new operator,
* require the baremetal-operator to create and delete HostOff/HostOn objects
  based on ``.Spec.Online`
* require HostOff objects to take precedence over HostOn,
* Hosts to be powered off if neither a HostOn or HostOff object exists,
* the operator to remove HostOff objects once the Host reaches a powered off
  state,
* machine-remediation to create/update HostOff objects and register a
  finaliser (in order to be able to delete the Node once the Host is off)
This would work, but is highly invasive, complex, and whether it provides any
tangible benefits is debatable.

```

Reply:
I don't think this does actually work - if multiple things need to create a HostOff resource and the resource name is unique then none of them know when to remove it. I guess you could always create it and delete it right away, and have everything just add finalisers (:screaming:), but I think that's just a more complicated way of doing the same thing as annotations.

In any event, this doesn't feel particularly declarative and I hate it enough that I don't feel the need to include it in the list of alternatives ;)

Reply:
The thinking was that the new controller (not machine-remediation) would delete the object once shutdown was complete. So m-r would create and add finizaliser

Reply:
And yes, i am not a fan either, but this is what alex came up with